File: Data_Structures/Stack/C/stack.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include "stack.h"


// Função para criar uma nova pilha
Pilha* inicializaPilha() {
    Pilha* p = (Pilha*)malloc(sizeof(Pilha));
    if (p == NULL) {
        printf("Erro: Não foi possível alocar memória para a nova pilha.\n");
        exit(1);
    }
    p->arr = (int*)malloc(sizeof(int));
    if (p->arr == NULL) {
        printf("Erro: Não foi possível alocar memória para a pilha.\n");
        exit(1);
    }
    p->topo = -1;
    p->qtd = 0;

    return p;
}

// Função para adicionar um novo elemento na pilha
void empilha(Pilha *p, int elemento) {
    p->qtd++;
    p->arr = (int*)realloc(p->arr, (p->qtd) * sizeof(int));
    if (p->arr == NULL) {
        printf("Erro: Não foi possível realocar memória para a pilha.\n");
        exit(1);
    }
    p->topo = elemento;
    p->arr[p->qtd - 1] = elemento;
}

// Função para remover o elemento do topo da pilha
int desempilha(Pilha *p) {
    if (p->topo == -1) {
        printf("Erro: A pilha está vazia.\n");
        exit(1);
    }

    p->qtd--;
    if(p->qtd == 0) {
        p->topo = -1;
        int elemento = p->arr[0];
        p->arr[0] = -1;
        return elemento;
    }

    p->arr = (int*)realloc(p->arr, (p->qtd) * sizeof(int));
    if (p->arr == NULL) {
        printf("Erro: Não foi possível realocar memória para a pilha.\n");
        exit(1);
    }
    int elemento = p->arr[p->qtd];
    p->topo = p->arr[p->qtd - 1];
    return elemento;
}

// Função para retornar o tamanho da pilha
int tamanhoPilha(Pilha *p) {
    return p->qtd;
}

// Função para verificar se a pilha está vazia
bool estaVazia(Pilha *p) {
    return p->qtd == 0;
}

// Função para retornar o elemento do topo da pilha
int topo(Pilha *p) {
    return p->topo;
}

File: Data_Structures/Stack/Rust/stack.rs
pub struct Stack {
    pub items: Vec<i32>,
    topo: Option<i32>,
    qtd: usize,
}

impl Stack{

    // Construtor da pilha
    pub fn new() -> Self {
        Stack {
            items: Vec::new(),
            topo: None,
            qtd: 0, 
        }
    }

    // Método para adicionar um novo elemento no topo da pilha
    pub fn empilha(&mut self, data: i32) {
        self.items.push(data);
        self.topo = Some(data);
        self.qtd += 1;
    }

    // Método para remover o elemento do topo da pilha
    pub fn desempilha(&mut self) -> Option<i32>{
        if self.qtd == 0 {
            return None; 
        }

        let elemento_removido = self.items.pop();
        self.topo = self.items.last().cloned();  
        self.qtd -= 1;

        elemento_removido 
    }

    // Método para retornar o elemento do topo da pilha
    pub fn topo_elemento(&self) -> Option<i32> {
        return self.topo;
    }

    // Método para retornar o tamanho da pilha
    pub fn tamanho(&self) -> i32 {
        return self.qtd as i32;
    }

    // Método para verificar se a pilha está vazia
    pub fn esta_vazia(&self) -> bool {
        return self.qtd == 0;
    }

}